File: python/asymmetric_parallel.py
import math
import time
import numpy as np
from multiprocessing import Pool, cpu_count
from asymmetric_sequential import generate_fractal_tree_asymmetric
from utils import print_header, print_params, print_result


def _worker(args):
    x, y, length, angle, left_ratio, right_ratio, left_angle_rad, right_angle_rad, min_length, depth = args
    branches = generate_fractal_tree_asymmetric(
        x, y, length, angle, left_ratio, right_ratio, left_angle_rad, right_angle_rad, min_length,
        start_depth=depth
    )
    return np.array(branches, dtype=np.float64)


def _build_tasks(x, y, length, angle, left_ratio, right_ratio, left_angle_rad, right_angle_rad,
                 min_length, depth, target_depth, upper_branches):
    if depth >= target_depth:
        return [(x, y, length, angle, left_ratio, right_ratio, left_angle_rad, right_angle_rad, min_length, depth)]

    end_x = x + length * math.cos(angle)
    end_y = y + length * math.sin(angle)
    upper_branches.append((x, y, end_x, end_y, depth))

    left = _build_tasks(end_x, end_y, length * left_ratio,  angle + left_angle_rad, 
                        left_ratio, right_ratio, left_angle_rad, right_angle_rad,
                        min_length, depth + 1, target_depth, upper_branches)

    right = _build_tasks(end_x, end_y, length * right_ratio, angle - right_angle_rad, 
                         left_ratio, right_ratio, left_angle_rad, right_angle_rad,
                         min_length, depth + 1, target_depth, upper_branches)
    return left + right


def run_parallel_asymmetric(trunk_length=100.0, left_ratio=0.67, right_ratio=0.57,
                             left_angle=35.0, right_angle=25.0,
                             min_length=0.01, num_processes=None, split_depth=None):

    if num_processes is None:
        num_processes = cpu_count()

    left_angle_rad  = math.radians(left_angle)
    right_angle_rad = math.radians(right_angle)
    if split_depth is None:
        split_depth = (num_processes * 4).bit_length() - 1

    print_header("Parallel Asymmetric (Python)")
    print_params(trunk_length, left_ratio, left_angle, min_length,
                 right_ratio=right_ratio, right_angle=right_angle,
                 cores=num_processes, split_depth=split_depth)

    start_time = time.perf_counter()

    start_x, start_y = 0, 0
    start_angle = math.pi / 2
    end_x = start_x + trunk_length * math.cos(start_angle)
    end_y = start_y + trunk_length * math.sin(start_angle)
    start_depth = 0

    upper_branches = [(start_x, start_y, end_x, end_y, start_depth)]

    left_child_len = trunk_length * left_ratio
    right_child_len = trunk_length * right_ratio
    left_child_angle_rad = start_angle + left_angle_rad
    right_child_angle_rad = start_angle - right_angle_rad

    # left subtree
    tasks = _build_tasks(end_x, end_y, 
                         left_child_len,left_child_angle_rad, 
                         left_ratio, right_ratio, 
                         left_angle_rad, right_angle_rad,
                         min_length, 1, split_depth, upper_branches)
    # right subtree
    tasks += _build_tasks(end_x, end_y, 
                          right_child_len, right_child_angle_rad,
                          left_ratio, right_ratio, 
                          left_angle_rad, right_angle_rad,
                          min_length, 1, split_depth, upper_branches)

    with Pool(processes=num_processes) as pool:
        results = pool.map(_worker, tasks)

    upper_array = np.array(upper_branches, dtype=np.float64)
    branches = np.concatenate([upper_array] + results)

    execution_time = time.perf_counter() - start_time

    total_branches = len(branches)
    max_depth = int(branches[:, 4].max()) if total_branches > 0 else 0
    print_result(execution_time, total_branches, max_depth)

    result = {
        'parameters': {
            'trunk_length': trunk_length,
            'left_ratio': left_ratio,
            'right_ratio': right_ratio,
            'left_angle': left_angle,
            'right_angle': right_angle,
            'min_length': min_length,
            'split_depth': split_depth,
            'num_processes': num_processes,
        },
        'execution_time': execution_time,
    }

    return result


if __name__ == "__main__":

    run_parallel_asymmetric(
        trunk_length=100.0,
        left_ratio=0.67,
        right_ratio=0.57,
        left_angle=35.0,
        right_angle=25.0,
        min_length=0.01,
    )

File: rust/src/asymmetric_parallel.rs
use fractal_tree::{
    generate_fractal_tree_asymmetric, collect_tasks_asymmetric,
    print_header, print_params, print_extra_param, print_result,
    Branch,
};
use rayon::prelude::*;
use std::f64::consts::PI;
use std::thread;
use std::time::Instant;
use std::error::Error;


pub fn run_parallel_asymmetric(
    trunk_length: f64,
    left_ratio: f64,
    right_ratio: f64,
    left_angle: f64,
    right_angle: f64,
    min_length: f64,
    num_threads: Option<usize>,
) -> Result<f64, Box<dyn Error>> {

    let num_threads = num_threads.unwrap_or_else(|| {
        thread::available_parallelism().map(|n| n.get()).unwrap_or(4)
    });

    let pool = rayon::ThreadPoolBuilder::new()
        .num_threads(num_threads)
        .build()?;

    let left_angle_rad  = left_angle.to_radians();
    let right_angle_rad = right_angle.to_radians();
    let split_depth = (num_threads * 4).next_power_of_two().ilog2() as usize;

    print_header("Parallel Asymmetric (Rust)");
    print_params(trunk_length, left_ratio, left_angle, min_length);
    print_extra_param("right_ratio", &right_ratio);
    print_extra_param("right_angle", &right_angle);
    print_extra_param("threads", &num_threads);
    print_extra_param("split_depth", &split_depth);

    let start = Instant::now();

    let mut upper_branches = vec![Branch { x1: 0.0, y1: 0.0, x2: 0.0, y2: trunk_length, depth: 0 }];

    // Left child of root uses left_ratio / left_angle; right child uses right_ratio / right_angle.
    let mut tasks = collect_tasks_asymmetric(
        &mut upper_branches,
        0.0, trunk_length, trunk_length * left_ratio, PI / 2.0 + left_angle_rad,
        left_ratio, right_ratio, left_angle_rad, right_angle_rad, 1, split_depth,
    );
    tasks.extend(collect_tasks_asymmetric(
        &mut upper_branches,
        0.0, trunk_length, trunk_length * right_ratio, PI / 2.0 - right_angle_rad,
        left_ratio, right_ratio, left_angle_rad, right_angle_rad, 1, split_depth,
    ));

    // Tasks have unequal sizes when left_ratio ≠ right_ratio — Rayon work-stealing
    // redistributes them dynamically, unlike Python's static pool.map split.
    let subtree_results: Vec<Vec<Branch>> = pool.install(|| {
        tasks.par_iter()
            .map(|task| generate_fractal_tree_asymmetric(
                task.x, task.y, task.length, task.angle,
                left_ratio, right_ratio, left_angle_rad, right_angle_rad,
                min_length, task.depth,
            ))
            .collect()
    });

    let execution_time = start.elapsed().as_secs_f64();

    let total_branches: usize = upper_branches.len()
        + subtree_results.iter().map(|s| s.len()).sum::<usize>();
    let max_depth = subtree_results.iter()
        .flat_map(|subtree| subtree.iter())
        .map(|b| b.depth)
        .max()
        .unwrap_or(0);

    print_result(execution_time, total_branches, max_depth);

    Ok(execution_time)
}


fn main() -> Result<(), Box<dyn Error>> {
    run_parallel_asymmetric(100.0, 0.67, 0.57, 35.0, 25.0, 0.01, None)?;
    Ok(())
}
